File: PyConfig/rise/scenario/Propagation.py
###############################################################################
# This file is part of openWNS (open Wireless Network Simulator)
# _____________________________________________________________________________
#
# Copyright (C) 2004-2007
# Chair of Communication Networks (ComNets)
# Kopernikusstr. 5, D-52074 Aachen, Germany
# phone: ++49-241-80-27910,
# fax: ++49-241-80-22242
# email: info@openwns.org
# www: http://www.openwns.org
# _____________________________________________________________________________
#
# openWNS is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License version 2 as published by the
# Free Software Foundation;
#
# openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################

from rise.scenario.IdRegistry import IdRegistry
import rise.scenario.Pathloss as Pathloss
import rise.scenario.Shadowing as Shadowing
import rise.scenario.FastFading as FastFading

class Configuration:
    """Holds a propagation configuration item.

    pathloss, shadowing and fastFading are intended to hold objects of the classes in
    rise.scenario.Pathloss, rise.scenario.Shadowing and rise.scenario.FastFading respectivly.
    """
    pathloss = None
    shadowing = None
    fastFading = None

    def __init__(self, pathloss = None, shadowing = None, fastFading = None):
        self.pathloss = pathloss
        self.shadowing = shadowing
        self.fastFading = fastFading


class NotConfigured(Exception):
    """Exception to be thrown if an unconfigured pair is requested in Propagation.getPair().
    """
    def __init__(self, pair, msgOutput = True):
        if (msgOutput):
            print
            print "Propagation not properly configured: Requested non-existent pair " + pair

class ChannelModelConfiguration():
    """
    Holds the configuration for a specific transceiver pair
    """

    def __init__(self, transceiverPair, channelmodel):
        self.transceiverPair = transceiverPair
        self.channelmodel = channelmodel

class Propagation:
    """Class to hold the propagation configuration.

    This class manages the propagation configuration for transmitter-receiver pairs. This
    encompasses the name to id mapping and the mapping of id pairs to Configuration objects.
    """
    __pairs = None
    __ids = None
    testRun = False

    def __init__(self):
        self.__pairs = {}
        self.__ids = IdRegistry()

    def knowsPairByName(self, transmitterName, receiverName):
        """ Check if a name pair has been configured.
        """
        if ((not self.knowsName(transmitterName)) or (not self.knowsName(receiverName))):
            return False
        return self.knowsPairById(self.__ids.getId(transmitterName),
                                  self.__ids.getId(receiverName))

    def knowsPairById(self, transmitterId, receiverId):
        """ Check if an id pair has been configured.
        """
        if ((not self.knowsId(transmitterId)) or (not self.knowsId(receiverId))):
            return False
        return (transmitterId, receiverId) in self.__pairs

    def setPair(self, transmitterName, receiverName):
        """ Configure a propagation pair.

        setPair() returns a Configuration instance. To configure pathloss, shadowing
        and fast fading, call it consecutivly:

        import rise.scenario.Pathloss as Pathloss
        import rise.scenario.Shadowing as Shadowing
        import rise.scenario.FastFading as FastFading
        propagationInstance.setPair('Foo', 'Bar').pathloss = Pathloss.FreeSpace()
        propagationInstance.setPair('Foo', 'Bar').shadowing = Shadowing.No()
        propagationInstance.setPair('Foo', 'Bar').fastFading = FastFading.No()
        """
        transmitterId = self.__ids.getId(transmitterName)
        receiverId = self.__ids.getId(receiverName)
        if (not self.knowsPairById(transmitterId, receiverId)):
            self.__pairs[transmitterId, receiverId] = Configuration()
        return self.__pairs[transmitterId, receiverId]

    def configure(self, configurations):
        """
        Configure pairs given by a list of ChannelModelConfigurations
        """
        for c in configurations:
            self.configurePair(c.transceiverPair[0], c.transceiverPair[1], c.channelmodel)

    def configurePair(self, transmitterName, receiverName, configuration):
        """ Configure a propagation pair.

        configurePair() takes a Configuration instance. To configure pathloss, shadowing
        and fast fading, you can prepare this configuration Object in advance.
        """
        transmitterId = self.__ids.getId(transmitterName)
        receiverId = self.__ids.getId(receiverName)
        self.__pairs[transmitterId, receiverId] = configuration

    def getPair(self, transmitterId, receiverId):
        """ Get the configuration for a pair.

        Returns Configuration for (transmitterId, receiverId) or raises
        NotConfigured if none was set.
        """
        try:
            return self.__pairs[transmitterId, receiverId]
        except KeyError:
            raise NotConfigured, (self.findName(transmitterId) + "->" + self.findName(receiverId), not self.testRun)

    def knowsName(self, name):
        return self.__ids.knowsName(name)

    def knowsId(self, id):
        return self.__ids.knowsId(id)

    def maxId(self):
        return self.__ids.len()

    def getId(self, name):
        return self.__ids.getId(name)

    def findId(self, name):
        return self.__ids.findId(name)

    def findName(self, id):
        return self.__ids.findName(id)

    def clear(self):
        self.__pairs = {}
        self.__ids = IdRegistry()

class DropInPropagation:
    __instance = None

    @staticmethod
    def getInstance():
        if (DropInPropagation.__instance == None):
            DropInPropagation.__instance = Propagation()
            DropInPropagation.__instance.setPair("DropIn", "DropIn").pathloss = Pathloss.FreeSpace()
            DropInPropagation.__instance.setPair("DropIn", "DropIn").shadowing = Shadowing.No()
            DropInPropagation.__instance.setPair("DropIn", "DropIn").fastFading = FastFading.No()
        return DropInPropagation.__instance

File: src/scenario/Propagation.cpp
/*******************************************************************************
 * This file is part of openWNS (open Wireless Network Simulator)
 * _____________________________________________________________________________
 *
 * Copyright (C) 2004-2007
 * Chair of Communication Networks (ComNets)
 * Kopernikusstr. 5, D-52074 Aachen, Germany
 * phone: ++49-241-80-27910,
 * fax: ++49-241-80-22242
 * email: info@openwns.org
 * www: http://www.openwns.org
 * _____________________________________________________________________________
 *
 * openWNS is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 2 as published by the
 * Free Software Foundation;
 *
 * openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <RISE/scenario/Propagation.hpp>
#include <RISE/scenario/pathloss/Pathloss.hpp>
#include <RISE/scenario/shadowing/Shadowing.hpp>
#include <RISE/scenario/fastfading/FastFading.hpp>

#include <WNS/pyconfig/helper/Functions.hpp>
#include <WNS/Ttos.hpp>
#include <WNS/Assure.hpp>

using namespace rise::scenario;

Propagation::NoModelError::NoModelError(const std::string& modelType,
					const std::string& transmitterName,
					const std::string& receiverName)
    : wns::Exception("No " + modelType + "-model configured for " + transmitterName + " -> " + receiverName)
{
}

Propagation::Propagation(const wns::pyconfig::View& config)
    : config(config),
    initialized_(false),
    logger_(wns::logger::Logger("RISE", "Propagation"))
{
    PathlossMatrix::SizeType sizes[2];
    sizes[0] = sizes[1] = config.get<std::size_t>("maxId()");
    pathlossMatrix = PathlossMatrix(sizes, NULL);
    shadowingMatrix = ShadowingMatrix(sizes, NULL);
    fastFadingMatrix = FastFadingMatrix(sizes, NULL);

    for (PathlossMatrix::SizeType i = 0; i < sizes[0]; ++i)
	{
		for (PathlossMatrix::SizeType j = 0; j < sizes[1]; ++j) 
		{
			if (config.get<bool>("knowsPairById(" + wns::Ttos(i) + ", " + wns::Ttos(j) + ")")) 
			{
				wns::pyconfig::View pairView = config.getView("getPair(" + wns::Ttos(i) + ", " + wns::Ttos(j) + ")");
				pathlossMatrix[i][j] = create<pathloss::Pathloss>(pairView.getView("pathloss"));
				shadowingMatrix[i][j] = create<shadowing::Shadowing>(pairView.getView("shadowing"));
				fastFadingMatrix[i][j] = create<fastfading::FastFading>(pairView.getView("fastFading"));

                pathlossMatrix[i][j]->setParentAndTypeIDs(this, i, j);
                shadowingMatrix[i][j]->setParentAndTypeIDs(this, i, j);
                fastFadingMatrix[i][j]->setParentAndTypeIDs(this, i, j);
			}
	    }
	}
}

Propagation::~Propagation()
{
}

void
Propagation::onWorldCreated()
{    
    if(initialized_)
        return;

    for (PathlossMatrix::SizeType i = 0; i < config.get<std::size_t>("maxId()"); ++i)
    {
        for (PathlossMatrix::SizeType j = 0; j < config.get<std::size_t>("maxId()"); ++j) 
    	{
			if (config.get<bool>("knowsPairById(" + wns::Ttos(i) + ", " + wns::Ttos(j) + ")")) 
			{
                MESSAGE_SINGLE(NORMAL, logger_, "Initializing models for [" 
                    << getName(i) << ";" << getName(j) << "]");

                pathlossMatrix[i][j]->onWorldCreated();
                shadowingMatrix[i][j]->onWorldCreated();            
                fastFadingMatrix[i][j]->onWorldCreated();
            }
        }
    }
    initialized_ = true;	
}

const pathloss::Pathloss&
Propagation::getPathlossModel(const IdType& transmitterId, const IdType& receiverId) const
{
    pathloss::Pathloss* model = pathlossMatrix[transmitterId][receiverId];
    if (model == NULL)
	throw NoModelError("pathloss", getName(transmitterId), getName(receiverId));
    return *model;
}

const shadowing::Shadowing&
Propagation::getShadowingModel(const IdType& transmitterId, const IdType& receiverId) const
{
    shadowing::Shadowing* model = shadowingMatrix[transmitterId][receiverId];
    if (model == NULL)
	throw NoModelError("shadowing", getName(transmitterId), getName(receiverId));
    return *model;
}

const fastfading::FastFading&
Propagation::getFastFadingModel(const IdType& transmitterId, const IdType& receiverId) const
{
    fastfading::FastFading* model = fastFadingMatrix[transmitterId][receiverId];
    if (model == NULL)
	throw NoModelError("shadowing", getName(transmitterId), getName(receiverId));
    return *model;
}

std::string
Propagation::getName(const IdType& id) const
{
    return config.get<std::string>("findName(" + wns::Ttos(id) + ")");
}

